### PR TITLE
feat(ilp): introduce less error-prone timestamp methods in Java ILP client

### PR DESCRIPTION
**Pull request by puzpuzpuz**
RFC: https://github.com/questdb/rfc/discussions/108

**Compatibility note**: this patch contains breaking API changes for the Java ILP client, as well as a server-breaking change. See the notes below.

### Client changes

**Compatibility note**: misleading `at(long)` and `timestampColumn(String, long)` methods are removed from `Sender` and replaced with the following safer alternatives:
```java
public interface Sender extends Closeable {

    void at(long timestamp, ChronoUnit unit);

    void at(Instant timestamp);

    Sender timestampColumn(CharSequence name, long value, ChronoUnit unit);

    Sender timestampColumn(CharSequence name, Instant value);
}
```

Internally, the new methods don't use the new timestamp suffixes described below, so the new client will be compatible with an older server.

### Server changes

Adds support for the following optional unit suffixes for timestamp and date columns in ILP/TCP messages (note: ILP/UDP is kept unchanged):
* `n` - nanosecond
* `t` - microsecond (old suffix, but now always interpreted as micros)
* `m` - millisecond

**Compatibility note**: previously, the `t` suffix was interpreted as epoch in micros for timestamp and date columns, but for the designated timestamp column it was interpreted as nanos. From now on, it's always interpreted as micros.

The suffixes are also supported for the designated timestamp column values, so the following ILP message is valid:
```ini
my_table ts_col=1630933921000000000n,date_col=1630933921ts 1630933921000000t
```

When a unit suffix is absent, the old conversion rules, e.g. `line.tcp.timestamp`, apply.

Also does the following:
* Marks `line.tcp.timestamp` and `line.udp.timestamp` as deprecated. These props are to be removed in future versions
* Adds timestamp examples to the Java client sample snippet used in the docs

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 138 / 139 (99.28%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpLexer.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpParser.java|62|62|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpTimestampAdapter.java|21|21|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMinuteTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineHourTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMicroTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/DefaultLineUdpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineException.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineNanoTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMilliTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineSecondTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/std/str/DirectByteCharSequence.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|8|8|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 137 / 138 (99.28%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpLexer.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpParser.java|62|62|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpTimestampAdapter.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMinuteTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineHourTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMicroTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/DefaultLineUdpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineException.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineNanoTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMilliTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineSecondTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/std/str/DirectByteCharSequence.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|8|8|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 137 / 138 (99.28%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpLexer.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpParser.java|62|62|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpTimestampAdapter.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMinuteTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineHourTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMicroTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/DefaultLineUdpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineException.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineNanoTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMilliTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineSecondTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/std/str/DirectByteCharSequence.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|8|8|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 158 / 167 (94.61%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/LineUdpSender.java|3|10|30.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpLexer.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cutlass/line/AbstractLineSender.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpParser.java|62|62|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpTimestampAdapter.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMinuteTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineHourTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMicroTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpSender.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/DefaultLineUdpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineException.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineNanoTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMilliTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineSecondTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/std/str/DirectByteCharSequence.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|8|8|100.00%|

File: core/src/main/java/io/questdb/cutlass/line/LineTcpSender.java [hunk 1]
```diff
@@ -74,12 +78,36 @@ public static LineTcpSender newSender(int ip, int port, int bufferCapacity) {
         }
     }
 
+    @Override
+    public final void at(long timestamp, ChronoUnit unit) {
+        put(' ').put(timestamp * unit.getDuration().toNanos() / 1000).put("tus");
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cutlass/line/LineTcpSender.java:83]: this is insane:

image

Could we check the sender latency before and after pls?

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 164 / 173 (94.80%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/LineUdpSender.java|3|10|30.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpLexer.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cutlass/line/AbstractLineSender.java|14|15|93.33%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpParser.java|62|62|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpTimestampAdapter.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMinuteTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineHourTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMicroTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpSender.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/DefaultLineUdpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineException.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineNanoTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMilliTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineSecondTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/std/str/DirectByteCharSequence.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|8|8|100.00%|

**Comment by amunra** [comment]
Can we consider doing this without mandatory server changes?
I'm introducing similar changes in the `c-questdb-client` which support passing in multiple types of timestamps, but ultimately do the conversion client-side to micros or nanos as appropriate: https://github.com/questdb/c-questdb-client/pull/42.

We should also deprecate the `line.tcp.timestamp` setting and ensure that columns (suffixed with `t` and therefore specified as micros) don't get interpreted as nanos when processing the designated timestamp. This would fix https://github.com/questdb/py-questdb-client/issues/47.

If we are to introduce new suffixes, these should be optional (maybe enabled via a client-side connection flag) to ensure compatibility of the client with a range of server versions.

I also only see scope to add a suffix for timestamps also for nanoseconds (i.e. `tns` or shorter I propose `tn`) with the intention of eventually adding support for a new `TIMESTAMP_NS` column type in QuestDB and retaining full precision of the timestamps we send from clients. This would be useful for the Python client especially, where dataframes hold timestamps as nanos almost exclusively for all columns already.

**Comment by bluestreak01** [comment]
>We should also deprecate the line.tcp.timestamp setting and ensure that ILP timestamp columns (currently always suffixed with t and therefore specified as micros) don't get interpreted as nanos when processing the designated timestamp. This would fix https://github.com/questdb/py-questdb-client/issues/47.

I agree, we should depricate first, remove later.

>If we are to introduce new suffixes, these should be optional (maybe enabled via a client-side connection flag) to ensure compatibility of the client with a range of server versions.

Suffixes are optional. In that old client will work with new server.

>I also only see scope to add a suffix in ILP timestamps for nanoseconds (i.e. tns or shorter I propose tn), and continuing to use t as the micros suffix, with the intention of eventually adding support for a new TIMESTAMP_NS column type in QuestDB. This would allow retaining full precision of the timestamps we receive from clients. This would be useful for the Python client especially, where dataframes hold timestamps as nanos almost exclusively for all columns already.

I would not send nanos from client to server _before_ nanos are implemented on the server. This leads to confusion.

The intention is to make the units of timestamps transparent for everyone:
- client API (it is up to client if to make client side conversion and always send micros or add suffix and have server convert)
- custom sender (people hand cranking ILP still), it is opportunity for them to avoid "maths" and send the original data marking what it is on the protocol
- having no feedback on ILP, we have to rely on server log to troubleshoot. Having suffix in the log helps understanding timestamp without also checking what's in the config.

We could converge on always sending micros afer all, but non-client users will be forced to do data conversion (and ultimately memorize that server accepts micros only). Vs sending explicit timestamp units and have code (and protocol) to be self explanatory.

**Comment by puzpuzpuz** [comment]
> I agree, we should depricate first, remove later.

I've marked both `line.*.timestamp` props as deprecated in ccd2d644693442e1f0125ef5d58012daf005023d. This means that they will be accepted and used, but there will be a warning printed to logs.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 166 / 175 (94.86%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/LineUdpSender.java|3|10|30.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpLexer.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cutlass/line/AbstractLineSender.java|14|15|93.33%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpParser.java|62|62|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpTimestampAdapter.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMinuteTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineHourTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMicroTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpSender.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/DefaultLineUdpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineException.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineNanoTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMilliTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineSecondTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/std/str/DirectByteCharSequence.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|8|8|100.00%|

**Comment by puzpuzpuz** [comment]
@amunra @bluestreak01 I've removed the suffices from the client impl and reduced the number of suffixes in 2c0ba0cd111b6fba2a3362b69a3d0f89da2ff041 and 55b91e019a0970a9c59ae31e4200c4763c6b2046

The PR should be ready for review. Note that there is a minor server-side incompatibility in handling `t` suffix + designated timestamps - see the description.

**Comment by amunra** [comment]
"but for the designated timestamp column it was interpreted as nanos."

That's a bummer.

Do any of our clients ever use the `t` suffix in the designated timestamp in ILP?

Is this grounds enough to pick a different suffix `u` maybe and deco the `t` suffix altogether?

(And I appreciate I'm probably bikeshedding at this stage)

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 126 / 140 (90.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/LineUdpSender.java|3|10|30.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMinuteTimestampAdapter.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cutlass/line/LineHourTimestampAdapter.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cutlass/line/LineSecondTimestampAdapter.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpLexer.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cutlass/line/AbstractLineSender.java|14|15|93.33%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpParser.java|33|35|94.29%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpTimestampAdapter.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMicroTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpSender.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/DefaultLineUdpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineException.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineNanoTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMilliTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/std/str/DirectByteCharSequence.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|8|8|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 126 / 140 (90.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/LineUdpSender.java|3|10|30.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMinuteTimestampAdapter.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cutlass/line/LineHourTimestampAdapter.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cutlass/line/LineSecondTimestampAdapter.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpLexer.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cutlass/line/AbstractLineSender.java|14|15|93.33%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpParser.java|33|35|94.29%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpTimestampAdapter.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMicroTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpSender.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/DefaultLineUdpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineException.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineNanoTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMilliTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/std/str/DirectByteCharSequence.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|8|8|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 126 / 140 (90.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/LineUdpSender.java|3|10|30.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMinuteTimestampAdapter.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cutlass/line/LineHourTimestampAdapter.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cutlass/line/LineSecondTimestampAdapter.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/LineUdpLexer.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cutlass/line/AbstractLineSender.java|14|15|93.33%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpParser.java|33|35|94.29%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpTimestampAdapter.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMicroTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineTcpSender.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/DefaultLineUdpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineException.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineNanoTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/LineMilliTimestampAdapter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/std/str/DirectByteCharSequence.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|8|8|100.00%|

**Comment by puzpuzpuz** [comment]
@bluestreak01 would be great to include this one into the upcoming release. Could you take a look?

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**